### DISCUSSION TITLE
Q: Why cannot box Span<T>?

### DISCUSSION THREAD

**Discussion by nathan130200** [dotnet/csharplang#6683] [Q&A] [answered]
I was creating static cache for types and its converters that uses `BinaryPrimitives`, `Span<T>` and `ReadOnlySpan<T>` but i cannot just box Span and ReadOnlySpan into object (base class?) for dynamic invoke function to converter T (primitives: int, short, long, uint, ushort, ulong).

```cs
delegate T Decoder<T>(ReadOnlySpan<byte> span);

var decoders = new Dictionary<Type, Delegate>()
{
  [typeof(int)] = new Decoder<int>(BinaryPrimitives.ReadInt32BitEndian)
};

var type = typeof(int);
var fn = decoders[type];
var buf = new byte[4]{1, 2, 3, 4};
var result = (int) fn.DynamicInvoke(buf.AsSpan()); // compile error: Cannot cast ReadOnlySpan<byte> to object

```

Don't work separate span into variable like:

```cs
var span = buf.AsSpan();
var result = (int) fn.DynamicInvoke(span);
```

Same happened with `Span<T>` i'm unable to use in DynamicInvoke for call function, and function for BinaryPrimitives only accept Span's

Why cannot just cast to base class

**Comment by HaloFour** [comment, marked as answer]
See: https://learn.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay

> These references are called interior pointers, and tracking them is a relatively expensive operation for the .NET runtime’s garbage collector. As such, the runtime constrains these refs to only live on the stack, as it provides an implicit low limit on the number of interior pointers that might be in existence.

> As a result, Span<T> instances can only live on the stack, not on the heap. This means you can’t box spans (and thus can’t use Span<T> with existing reflection invoke APIs, for example, as they require boxing). It means you can’t have Span<T> fields in classes, or even in non-ref-like structs. It means you can’t use spans in places where they might implicitly become fields on classes, for instance by capturing them into lambdas or as locals in async methods or iterators (as those “locals” may end up being fields on the compiler-generated state machines.) It also means you can’t use Span<T> as a generic argument, as instances of that type argument could end up getting boxed or otherwise stored to the heap (and there’s currently no “where T : ref struct” constraint available).

**Reply by nathan130200** [reply]
But how to proceed with in BinaryPrimitives utility class? They should implement an variant for `Memory<T>` and `ReadOnlyMemory<T>` also.

Because even if i make runtime an generic method instance, when i call `Invoke` on it, will required `object` breaking Span's boxing rule.

**Reply by HaloFour** [reply]
You can't use `DynamicInvoke` with those types, you'll have to use `Invoke` directly on a proper delegate type.

**Reply by nathan130200** [reply]
I cant still use Invoke. This still requires Span to be boxed to object before call.

**Reply by nathan130200** [reply]
Cannot implicit cast `ReadOnlySpan<byte>` to object.

![image](https://user-images.githubusercontent.com/14365254/201123677-9892bae0-fb5d-4ead-a451-691b63e36534.png)

**Reply by HaloFour** [reply]
Yes, these reflective calls are illegal.  You need to cast the delegate instance back to a proper delegate type and invoke it directly:

```cs
static T ReadValue<T>(ReadOnlySpan<byte> span) {
    var fn = (Decoder<T>) decoders[typeof(T)];
    return fn(span);
}
```

**Reply by nathan130200** [reply]
Oh my, really :joy: I hadn't even seen that. In my mind I was still doing reflection to access the function. Thanks for help.
